File: src/main/java/jokatu/util/Json.java
package jokatu.util;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.core.convert.converter.Converter;
import org.springframework.core.convert.converter.ConverterFactory;

import java.io.IOException;

public class Json {

	public static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

	public static String serialise(Object object) throws JsonProcessingException {
		return OBJECT_MAPPER.writeValueAsString(object);
	}

	public static class JacksonConverterFactory implements ConverterFactory<String, Object> {

		@Override
		public <T> Converter<String, T> getConverter(Class<T> targetType) {
			return source -> {
				try {
					return OBJECT_MAPPER.readValue(source.getBytes(), targetType);
				} catch (IOException e) {
					throw new RuntimeException(e);
				}
			};
		}
	}
}

File: src/main/kotlin/jokatu/util/Json.kt
package jokatu.util

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.core.convert.converter.Converter
import org.springframework.core.convert.converter.ConverterFactory
import java.io.IOException
import java.util.*

/**
 * A utility class for handling JSON.
 */
object Json {

	private val OBJECT_MAPPER = ObjectMapper()

	@Throws(JsonProcessingException::class)
	// This is called from the Groovy Templates.
	fun serialise(`object`: Any?): String {
		return OBJECT_MAPPER.writeValueAsString(`object`)
	}

	@Throws(JsonProcessingException::class)
	fun serialiseAndEscape(`object`: Any?): String {
		return serialise(`object`).replace("'".toRegex(), "&#39;")
	}

	@Throws(IOException::class)
	// This can be called from the Groovy Templates.
	fun deserialise(json: String): Map<String, Any?> {
		val typeRef = OBJECT_MAPPER.typeFactory.constructMapType(HashMap::class.java, String::class.java, Any::class.java)
		return OBJECT_MAPPER.readValue(json, typeRef)
	}

	class JacksonConverterFactory : ConverterFactory<String, Any?> {
		override fun <T> getConverter(targetType: Class<T>): Converter<String, T> {
			return Converter { source ->
				try {
					return@Converter OBJECT_MAPPER.readValue<T>(source.toByteArray(), targetType)
				} catch (e: IOException) {
					throw RuntimeException(e)
				}
			}
		}
	}
}
